lib/surface/compiler.ex

```diff
@@ -324,6 +325,72 @@ defmodule Surface.Compiler do
      }}
   end
 
+  defp convert_node_to_ast(
+         :for_else,
+         {"#else", _attributes, children, node_meta},
+         compile_meta
+       ) do
+    meta = Helpers.to_meta(node_meta, compile_meta)
+
+    {:ok,
+     %AST.Container{
+       children: to_ast(children, compile_meta),
+       meta: meta,
+       directives: []
+     }}
+  end
+
+  defp convert_node_to_ast(
+         :for_else,
+         {_name, attributes, children, node_meta},
+         compile_meta
+       ) do
+    meta = Helpers.to_meta(node_meta, compile_meta)
+    default = %AST.AttributeExpr{value: false, original: "", meta: meta}
+    generator = attribute_value_as_ast(attributes, "each", :generator, default, compile_meta)
+
+    [for_children, else_children] =
+      case children do
+        [{:default, [], default, _}, {"#else", _, _, _} = else_block] ->
+          [default, [else_block]]
+
+        [{:default, [], default, _}] ->
+          [default, []]
+
+        children ->
+          [children, []]
+      end
+
+    for_ast = %AST.For{
+      generator: generator,
+      children: to_ast(for_children, compile_meta),
+      meta: meta
+    }
+
+    if else_children == [] do
+      {:ok, for_ast}
+    else
+      [{_, _, [{_, _, _}, value]}] = generator.value
+
+      condition = %AST.AttributeExpr{
+        original: "",
+        value:
+          quote generated: true do
+            unquote(value) != []
+          end,
+        meta: compile_meta
```

Comment:
So this works but I'm almost ashamed of having written it 😅 

@Malian @lnr0626 @msaraiva I need smarter people than me to enlighten me

Reply:
@miguel-s I would introduce a new `else` attribute in the `AST.For` instead of wrapping the `For` in the `If`. The main reason is to have an AST as close as possible of the user code.

We could handle the two cases in the `EExEngine`.

Another solution is to wait that for #378 :sweat_smile:

Reply:
@miguel-s I don't think that will work when using filters, like:

```
<#for {i <- items, items in [...]}>
  ...
</#for>
```

The condition of the `If` must test against the result of the `for`. We need to generate something like:

```elixir
for {i <- items, items in [...]} do
  ...
end |> case do
  [] ->
    # The else block
  result ->
    result
end
```

However, I'm not sure if creating this variable would mess with diff tracking inside the `for` block.